### ISSUE TITLE
Aesara as an array backend in Xarray

### ISSUE THREAD

**Issue by jhamman** [pydata/xarray#7515]
### Is your feature request related to a problem?

I recently learned about a meta-tensor library called [Aesara](https://aesara.readthedocs.io/) which got me wondering if it would be a good array backend for Xarray.

> Aesara is a Python library that allows you to define, optimize/rewrite, and evaluate mathematical expressions involving multi-dimensional arrays efficiently. It is composed of different parts:
> - Symbolic representation of mathematical operations on arrays
> - Speed and stability optimization
> - Efficient symbolic differentiation
> - Powerful rewrite system to programmatically modify your models
> - Extendable backends. Aesara currently compiles to C, Jax and Numba.

![image](https://user-images.githubusercontent.com/2443309/217439615-724db5f3-80c1-4577-ac95-620b2b27bf72.png)

xref: https://github.com/aesara-devs/aesara/issues/352, @OriolAbril, @twiecki

Has anyone looked into this yet?

**Comment by twiecki** [comment]
@jhamman Yes, I think this is an interesting discussion to have. We actually want to add named dimensions and coordinates to PyTensor (a fork of aesara): https://github.com/pymc-devs/pytensor, adding this capability was one of the reasons for forking, as PyMC already has some support built-in (coords and dims) and uses xarray in various places. Tighter integration between the two is a great goal.

**Comment by brandonwillard** [comment]
> @jhamman Yes, I think this is an interesting discussion to have. We actually want to add named dimensions and coordinates to PyTensor (a fork of aesara): https://github.com/pymc-devs/pytensor, adding this capability was one of the reasons for forking, as PyMC already has some support built-in (coords and dims) and uses xarray in various places. Tighter integration between the two is a great goal.

Hi, I'm the creator and maintainer of Aesara and we _are_ interested in discussing named dimensions and arrays.

We don't know why @twiecki's and his PyMC Labs group has been attempting to speak for us on this&mdash;or any other&mdash;matter.  As you can see, the original issue in Aesara is still open, and we have **not** said that the feature cannot be implemented in some way.

**Comment by rlouf** [comment]
Author of the little diagram and core member of Aesara here 🙂 If you want to know a little more about where the project is headed you can consult the project’s [Mission statement](https://github.com/aesara-devs/aesara/blob/main/doc/mission.rst). Happy to chat, you can find my contact info on my GitHub profile!

**Comment by TomNicholas** [comment]
(I was already typing this when Ryan posted so I'll finish anyway :sweat_smile:)

To clarify, what @jhamman is suggesting in this specific issue is **xarray wrapping aesara**, as opposed to **aesara wrapping xarray**. 

Both of these goals would be great, but in xarray we would particularly love to be able to wrap aesara because:
- We already have a large community of users who use xarray objects as their top-level interface, 
- It continues a [larger project](https://github.com/pydata/xarray/projects/2) of xarray being generalized to wrap any numpy-like "duck array".
- Our [ultimate goal](https://github.com/pydata/xarray/issues/6807) is for xarray users to be able to seamless switch out their computing backend and find which library gives them the best performance without changing the rest of their high-level code.

For xarray to wrap aesara aesara needs to provide a numpy-like API, ideally conforming to the [python array api standard](https://data-apis.org/array-api/latest/). If aesara already does this then we should try out the wrapping right now!

If you're interested in this topic I invite you to drop in to a meeting of the [Pangeo working group on distributed arrays](https://discourse.pangeo.io/t/new-working-group-for-distributed-array-computing/2734)! We have so far had talks from distributed computing libraries including Arkouda, Ramba, and cubed, all of which we are hoping to support as compute backends.

---

If anyone is also separately interested in using xarray inside PyTensor / aesara then that's awesome, but we should try to track efforts in that direction on a different issue to keep this distinction clear. We plan to better support that direction of wrapping soon by fully exposing our (semi-private internal currently) lightweight `Variable` class.

**Comment by jhamman** [comment]
Thanks all for the discussion. Welcome @brandonwillard, @twiecki, and @rlouf to the Xarray project. And thanks to @rabernat and @TomNicholas for helping orient the conversation.

As a next step, I think it would be fun to try putting an Aesara array into Xarray and see how it goes. In our experience, this process inevitably brings up a few issues where interaction between Xarray and Aesara developers is fruitful.

**Comment by rlouf** [comment]
Aesara has a NumPy-like API, and we always planned to conform to the python array api standard (see https://github.com/aesara-devs/aesara/issues/729 for instance). This would be the impetus for us to do it now.

**Comment by jhamman** [comment]
Here's a small example to get things started:

```python
x = aesara.shared(np.random.standard_normal((3, 4)))
xda = xr.DataArray(x, dims=('x', 'y'))
```
this currently returns the following error:
```python traceback
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[138], line 2
      1 x = aesara.shared(np.random.standard_normal((3, 4)))
----> 2 xda = xr.DataArray(x, dims=('x', 'y'))

File ~/miniforge3/envs/demo-env/lib/python3.10/site-packages/xarray/core/dataarray.py:428, in DataArray.__init__(self, data, coords, dims, name, attrs, indexes, fastpath)
    426 data = _check_data_shape(data, coords, dims)
    427 data = as_compatible_data(data)
--> 428 coords, dims = _infer_coords_and_dims(data.shape, coords, dims)
    429 variable = Variable(dims, data, attrs, fastpath=True)
    430 indexes, coords = _create_indexes_from_coords(coords)

File ~/miniforge3/envs/demo-env/lib/python3.10/site-packages/xarray/core/dataarray.py:142, in _infer_coords_and_dims(shape, coords, dims)
    140     dims = tuple(dims)
    141 elif len(dims) != len(shape):
--> 142     raise ValueError(
    143         "different number of dimensions on data "
    144         f"and dims: {len(shape)} vs {len(dims)}"
    145     )
    146 else:
    147     for d in dims:

ValueError: different number of dimensions on data and dims: 0 vs 2
```

This tells me there is a bit of work to do at the core of Aesara's numpy compatibility. Xarray will make frequent references to attributes like `data.shape`, `data.ndim`, etc expecting to get numpy-like results.

**Comment by brandonwillard** [comment]
> This tells me there is a bit of work to do at the core of Aesara's numpy compatibility. Xarray will make frequent references to attributes like `data.shape`, `data.ndim`, etc expecting to get numpy-like results.

We'll fix the compatibility issues, but we first need to understand what the expectations on something like `data.shape` should be in these circumstances.

For a shared variable, it's always possible to get a value for `data.shape` by referencing the underlying data, but the reason we don't do that by default is&mdash;in part&mdash;due to the fact that shared variables can be updated with values that have different shapes (but the same dtypes and number of dimensions).

**Comment by TomNicholas** [comment]
> We'll fix the compatibility issues, but we first need to understand what the expectations on something like data.shape should be in these circumstances.

At a minimum, xarray expects `.shape`, `.ndim` and `.dtype` to always be defined. On top of that there are extra expectations about slicing and broadcasting changing shape in the same ways as it does for numpy arrays. (@keewis correct me if I've mis-stated this or missed something important here!)

> For a shared variable, it's always possible to get a value for data.shape by referencing the underlying data, but the reason we don't do that by default is—in part—due to the fact that shared variables can be updated with values that have different shapes (but the same dtypes and number of dimensions).

This sounds a bit similar to discussions we have been having about [wrapping ragged arrays in xarray](https://github.com/pydata/xarray/issues/4285), for which there are multiple ways you might choose to define the shape.

The simplest way to guarantee that aesara can be wrapped by xarray is if aesara conformed to the array API standard, and they have [a test suite](https://github.com/pydata/xarray/issues/6894#issuecomment-1255644395) you can use to check that. We are also [working on our own testing framework](https://github.com/pydata/xarray/issues/6894) that duck-typed array libraries like aesara could import to quickly test integration with xarray.

**Comment by TomNicholas** [comment]
Just wanted to drop in and remind people interested in this that we hold a bi-weekly [pangeo working group for distributed array computing](https://discourse.pangeo.io/t/new-working-group-for-distributed-array-computing/2734), which is the perfect place to come and ask about any questions over zoom! I'll be there at 1pm EST today.
